rfcs/text/0015_nodejs_clustering.md

```diff
@@ -0,0 +1,526 @@
+  /**
+   * Returns true if the current worker has been elected as the main one. In non-clustered mode, will return true
+   */
+  isMainWorker: () => boolean;
+}
+```
+
+## Cross-worker communication
+
+For some of our changes (such as the /status API, see below), we will need some kind of cross-worker communication. This 
+will need to pass through the coordinator, which will also serve as an 'event bus', or IPC forwarder.
+
+This IPC API will be exposed from the clustering service.
+
+```ts
+export interface ClusteringServiceSetup {
+  // [...]
+  broadcast: (type: string, payload?: ClusterMessagePayload, options?: BroadcastOptions) => void;
+  addMessageHandler: (type: string, handler: MessageHandler) => MessageHandlerUnsubscribeFn;
+}
+```
+
+Notes:
+-  to reduce clustered and non-clustered mode divergence, in non-clustered mode, these APIs would just be no-ops. 
+   It will avoid to force (most) code to check in which mode Kibana is running before calling them.
+-  We could eventually use an Observable pattern instead of a handler pattern to subscribe to messages.
+
+## Executing code on a single worker
+
+In some scenarios, we would like to have parts of the code executed only from a single process. 
+
+The SO migration would be a good example: 
+we don't need to have each worker try to perform the migration, and we'd prefer to have one performing/trying the migration, 
+and the others wait for it. Due to the architecture, we can't have the coordinator perform such single-process jobs, 
+as it doesn't actually run a Kibana server.
+
+There are various ways to address such use-cases. What seems to be the best compromise right now would be the concept of 
+'main worker'. The coordinator would arbitrary elect a worker as the 'main' one at startup. The clustering service would then expose 
+an API to let workers identify themselves as main or not.
+
+```ts
+export interface ClusteringServiceSetup {
+  // [...]
+  isMainWorker: () => boolean;
+}
+```
+
+Notes:
+- in non-clustered mode, `isMainWorker` would always return true, to reduce the divergence between clustered and 
+  non-clustered modes.
+
+### Example: SO Migration
+
+To take the example of SO migration, the `KibanaMigrator.runMigrations` implementation could change to
+(naive implementation, the function is supposed to return a promise here, did not include that for simplicity):
+
+```ts
+runMigration() {
+   if(clustering.isMainWorker()) {
+     this.runMigrationsInternal().then((result) => {
+        applyMigrationState(result);
+        clustering.broadcast('migration-complete', { payload: result }, { persist: true }); // persist: true will send message even if subscriber subscribe after the message was actually sent
+      })
+   }
+   else {
+     const unsubscribe = clustering.addMessageHandler('migration-complete', ({ payload: result }) => {
+       applyMigrationState(result);
+       unsubscribe();
+     });
```

Comment:
We may want some guard rails around this message handling API. For example, I think it would be best if we segmented these messages so that plugins can only read messages that come from the same plugin to avoid any implicit cross-plugin APIs.